src/skmultiflow/anomaly_detection/half_space_trees.py

```diff
@@ -0,0 +1,548 @@
+        return normalize(y_proba_mean, norm='l1')
+
+    def initialise_work_space(self):
+        """
+        Initialises work spaces.
+        For every dimension in the feature space, creates a minimum and a maximum work range.
+        """
+        for i in range(self.dimensions):
+            sq = random.uniform(0, 1)
+            min_element = sq - 2 * max(sq, 1 - sq)
+            max_element = sq + 2 * max(sq, 1 - sq)
+            try:
+                self.min[i] = min_element
+            except IndexError:
+                self.min.append(min_element)
+            try:
+                self.max[i] = max_element
+            except IndexError:
+                self.max.append(max_element)
+
+    def build_trees(self):
+        """
+        Initialises ensemble.
+
+        """
+        for i in range(self.n_estimators):
+            self.initialise_work_space()
+            tree = HalfSpaceTree(self.depth, self.dimensions, self.size_limit)
+            tree.root = tree.build_tree(self.min, self.max)
+            self.ensemble.append(tree)
+
+    def update_mass(self, X, boolean):
+        """
+        Populates mass profiles for every tree defined in the ensemble.
+
+        Parameters
+        ----------
+        X: numpy.ndarray of shape (1, n_features)
+            Instance attributes.
+
+        boolean: boolean, True or False
+            True to update reference mass, False to update latest mass
+        """
+        for i in range(self.n_estimators):
+            self.ensemble[i].update_mass(X, boolean)
+
+    def update_models(self):
+        """
+        Updates the mass profile of every tree in the ensemble.
+
+        """
+        for i in range(self.n_estimators):
+            tree = self.ensemble[i]
+            tree.update_model(tree.root)
+
+    def set_is_learning_phase_on(self, boolean):
+        """
+        Sets learning phase in each tree defined in the ensemble.
+
+        Parameters
+        ----------
+        boolean: Boolean
+
+        """
+        self.is_learning_phase_on = boolean
+        for i in range(self.n_estimators):
+            self.ensemble[i].is_learning_phase_on = boolean
+
+
+class HalfSpaceTree:
+
+    def __init__(self, max_depth, dimensions, size_limit):
+        """
+        Half Space Tree
+
+        Parameters
+            ----------
+        dimensions: int, required
+            The dimensionality of the stream.
+
+        max_depth: int, optional (default=15)
+            The maximum depth of the trees in the ensemble.
+            'maxDepth' in the original paper.
+
+        size_limit: int, optional (default=50)
+            The minimum mass required in a node (as a fraction of the window size) to calculate the anomaly score.
+            'sizeLimit' in the original paper.
+        """
+        super().__init__()
+        self.max_depth = max_depth
+        self.dimensions = dimensions
+        self.root = HalfSpaceTreeNode()
```

Comment:
Why not call `build_tree(...)` here? This way we avoid having two calls in the parent class:
 ```python
tree = HalfSpaceTree(self.depth, self.dimensions, self.size_limit)
tree.root = tree.build_tree(self.min, self.max)
```
This would require a slight change in the class declaration, however, the result is closer to the MOA implementation.

Reply:
Your question sums up a bit to what you have cited in your first comment:

> My first question is about the 3 classes (HalfSpaceTrees, HalfSpaceTree and HalfSpaceTreeNode) declared instead of the 2 in the MOA implementation (HSTrees and HSTreeNode). Why is this? It seems to me that this adds complexity when reading the code and differentiates both implementations. 

I was thinking that having the HalfSpaceTree class makes the readability a bit easier. Maybe I am thinking way too OOP but since it is a ensemble method, instances of the ensemble should be able to decide individually also. For my point of view, a Half Space Tree should be the candidate for those functions not a Half Space Tree Node. Maybe you can explain why you were thinking like that.

Reply:
I agree with your reasoning. I assumed it what you mentioned, but asked for confirmation. :+1: for readibility.

My comment here is about instantiating a Half--space tree on just one line instead of two:

```python
tree = HalfSpaceTree(self.depth, self.dimensions, self.size_limit, self.min, self.max)
```

With a call to `build_tree` from the `__init__` method.

Reply:
Yes, I understand now, it makes it easier to read.

Reply:
I have integrated some changes, is it possible to resolve some of the conversation? It started to be a bit hard to navigate.